File: sdks/python3/examples/chain_test_node.py
from __future__ import annotations
from datetime import datetime, timezone
from typing import Any, Dict

from blok.node.node_handler import NodeHandler
from blok.types.context import Context


class ChainTestNode(NodeHandler):
    """Cross-runtime chain test node for integration testing.

    Reads a chain array from the request body, appends a python3 entry,
    and returns the updated chain -- proving data flows between languages.
    """

    def execute(self, ctx: Context, config: Dict[str, Any]) -> Any:
        body = ctx.request.body_map()

        # Read existing chain — gRPC inputs first (carried on `node.config`),
        # HTTP body fallback (legacy wire shape where the runner mapped
        # resolvedInputs → request.body). Dual-read keeps the
        # cross-runtime-chain demo working over both transports during
        # the §11 deprecation window.
        chain: list = []
        if isinstance(config.get("chain"), list):
            chain = list(config["chain"])
        elif body and isinstance(body.get("chain"), list):
            chain = list(body["chain"])

        # Read origin — same dual-read.
        origin = "unknown"
        if isinstance(config.get("origin"), str) and config["origin"]:
            origin = config["origin"]
        elif body and isinstance(body.get("origin"), str):
            origin = body["origin"]

        # Append this language's entry
        entry = {
            "language": "python3",
            "order": len(chain) + 1,
            "timestamp": datetime.now(timezone.utc).isoformat(),
        }
        chain.append(entry)

        # Store in context vars
        ctx.set_var("chain", chain)

        return {
            "chain": chain,
            "origin": origin,
        }

File: sdks/java/src/main/java/com/blok/blok/nodes/ChainTestNode.java
package com.blok.blok.nodes;

import com.blok.blok.node.NodeHandler;
import com.blok.blok.types.Context;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * ChainTestNode is used in cross-runtime integration tests.
 * It reads a chain array from the request body, appends its own entry,
 * and returns the updated chain — proving data flows between languages.
 */
public class ChainTestNode implements NodeHandler {

    @Override
    @SuppressWarnings("unchecked")
    public Object execute(Context ctx, Map<String, Object> config) throws Exception {
        Map<String, Object> body = null;
        if (ctx.getRequest() != null && ctx.getRequest().getBody() instanceof Map) {
            body = (Map<String, Object>) ctx.getRequest().getBody();
        }

        // Read existing chain — gRPC inputs first (carried on
        // `node.config`), HTTP body fallback (legacy wire shape where
        // the runner mapped resolvedInputs → request.body). Dual-read
        // keeps the cross-runtime-chain demo working over both
        // transports during the §11 deprecation window.
        List<Object> chain = new ArrayList<>();
        if (config != null && config.get("chain") instanceof List) {
            chain = new ArrayList<>((List<Object>) config.get("chain"));
        } else if (body != null && body.get("chain") instanceof List) {
            chain = new ArrayList<>((List<Object>) body.get("chain"));
        }

        // Read origin — same dual-read.
        String origin = "unknown";
        if (config != null && config.get("origin") instanceof String s && !s.isEmpty()) {
            origin = s;
        } else if (body != null && body.get("origin") instanceof String s && !s.isEmpty()) {
            origin = s;
        }

        // Append this language's entry
        Map<String, Object> entry = new HashMap<>();
        entry.put("language", "java");
        entry.put("order", chain.size() + 1);
        entry.put("timestamp", Instant.now().toString());
        chain.add(entry);

        // Store in context vars
        if (ctx.getVars() != null) {
            ctx.getVars().put("chain", chain);
        }

        // Return updated chain
        Map<String, Object> result = new HashMap<>();
        result.put("chain", chain);
        result.put("origin", origin);
        return result;
    }
}
